src/main/app/document/DocumentManagementController.ts

```diff
@@ -303,22 +230,163 @@ export class DocumentManagerController extends PostController<AnyObject> {
+  ) {
+    let uid = '';
+    let documentToGet = '';
+    let ele = '';
+    const document_filename = req.session.userCase[`${element}`]?.document_filename;
+
+    if (element !== null && element !== undefined) {
+      ele = element[0];
+      if (!req.session.userCase[`${ele}`]?.document_binary_url) {
+        throw new Error('binary url is not found for ' + document_filename);
+      }
+      documentToGet = req.session.userCase[`${ele}`]?.document_binary_url;
+      uid = this.getUID(documentToGet);
+
+      if (flag !== null || flag !== undefined) {
+        flag = YesOrNo.YES;
+      }
+    }
+    return uid;
+  }
+
+  private async setFlagViewed(
+    req: AppRequest<Partial<CaseWithId>>,
+    caseReference: string,
+    client: CosApiClient,
+    loggedInCitizen: UserDetails,
+    flag: string
+  ) {
+    let isFlagViewed;
+    let skipCallToSaveState: boolean;
+    req?.session?.userCase.respondents?.forEach((respondent: Respondent) => {
+      const cvIsApplicationViewed = respondent?.value?.response?.citizenFlags?.isApplicationViewed;
+      const cvIsAllegationOfHarmViewed = respondent?.value?.response?.citizenFlags?.isAllegationOfHarmViewed;
+
+      if (respondent?.value?.user?.idamId === req.session?.user.id) {
+        if (flag === DownloadFileFieldFlag.IS_APPLICATION_VIEWED && cvIsApplicationViewed === YesOrNo.YES) {
+          skipCallToSaveState = true;
+        } else if (
+          flag === DownloadFileFieldFlag.IS_ALLEGATION_OF_HARM_VIEWED &&
+          cvIsAllegationOfHarmViewed === YesOrNo.YES
+        ) {
+          skipCallToSaveState = true;
+        }
+
+        if (!skipCallToSaveState) {
+          this.setCaseDataCitizenFlags(flag, cvIsAllegationOfHarmViewed, cvIsApplicationViewed, respondent);
+          isFlagViewed = YesOrNo.YES;
+        }
+      }
+    });
+    if (isFlagViewed) {
+      const data = toApiFormat(req?.session?.userCase);
+      data.id = caseReference;
+      const updatedCaseDataFromCos = await client.updateCase(
+        loggedInCitizen,
+        caseReference as string,
+        data,
+        'citizen-internal-case-update'
+      );
+      req.session.userCase = updatedCaseDataFromCos;
+    }
+  }
+
+  private setCaseDataCitizenFlags(
+    flag: string,
+    cvIsAllegationOfHarmViewed: string | undefined,
+    cvIsApplicationViewed: string | undefined,
+    respondent: Respondent
+  ) {
+    if (flag === DownloadFileFieldFlag.IS_APPLICATION_VIEWED && respondent?.value?.response?.citizenFlags) {
+      if (
+        cvIsAllegationOfHarmViewed === null ||
+        cvIsAllegationOfHarmViewed === undefined ||
+        cvIsAllegationOfHarmViewed === 'No'
+      ) {
+        respondent.value.response.citizenFlags = {
+          isAllegationOfHarmViewed: 'No',
+          isApplicationViewed: 'Yes',
+        };
+      } else {
+        respondent.value.response.citizenFlags = {
+          isAllegationOfHarmViewed: 'Yes',
+          isApplicationViewed: 'Yes',
+        };
+      }
```

Comment:
Plz update only one field which is required

Reply:
code changes done